cli/tests/test_operations.rs

```diff
@@ -508,6 +508,84 @@ fn test_op_abandon_without_updating_working_copy() {
     "###);
 }
 
+#[test]
+fn test_op_abandon_multiple_heads() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "init", "repo"]);
+    let repo_path = test_env.env_root().join("repo");
+
+    // Create 1 base operation + 2 operations to be diverged.
+    test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "commit 1"]);
+    test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "commit 2"]);
+    test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "commit 3"]);
+    let stdout = test_env.jj_cmd_success(
+        &repo_path,
+        &["op", "log", "--no-graph", r#"-Tid.short() ++ "\n""#],
+    );
+    let (head_op_id, prev_op_id) = stdout.lines().next_tuple().unwrap();
+    insta::assert_snapshot!(head_op_id, @"cd2b4690faf2");
+    insta::assert_snapshot!(prev_op_id, @"c2878c428b1c");
+
+    // Create 1 other concurrent operation.
+    test_env.jj_cmd_ok(&repo_path, &["commit", "--at-op=@--", "-m", "commit 4"]);
+
+    // Can't resolve operation relative to @.
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["op", "abandon", "@-"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: The "@" expression resolved to more than one operation
+    Hint: Try specifying one of the operations by ID: cd2b4690faf2, 603773cdd351
+    "###);
+    let (_, other_head_op_id) = stderr.trim_end().rsplit_once(", ").unwrap();
+    insta::assert_snapshot!(other_head_op_id, @"603773cdd351");
+    assert_ne!(head_op_id, other_head_op_id);
+
+    // Can't abandon one of the head operations.
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["op", "abandon", head_op_id]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: Cannot abandon the current operation cd2b4690faf2
+    Hint: Run `jj undo` to revert the current operation, then use `jj op abandon`
+    "###);
+
+    // Can't abandon the other head operation.
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["op", "abandon", other_head_op_id]);
+    insta::assert_snapshot!(stderr, @r###"
+    Error: Cannot abandon the current operation 603773cdd351
+    Hint: Run `jj undo` to revert the current operation, then use `jj op abandon`
+    "###);
+
+    // Can abandon the operation in one branch, and the other head should be
```

Comment:
What do you mean by this? It looks like we abandon the operation *before* the branch point. And what does "the other head should be unchanged" mean?

Reply:
Suppose the following graph, `prev_op` is abandoned and `head_op` is reparented. `other_head_op` is also reparented (to the same base operation), but that should be noop.
```
  o other_head_op
  |
o | head_op
| |
o | prev_op
|/
o
```

Perhaps, ", and the other head should be unchanged." can be just removed. This is the test for "would crash if we attempted to remap the unchanged op."

Reply:
Updated to
```
    // Can abandon the operation which is not an ancestor of the other head.
    // This would crash if we attempted to remap the unchanged op in the op
    // heads store.
```